lib/improver/spotdata_new/neighbour_finding.py

```diff
@@ -0,0 +1,517 @@
+                in the target coordinate system, shaped as (n_sites, 2). The
+                z coordinate column is excluded from the return.
+        """
+        target_coordinate_system = cube.coord_system().as_cartopy_crs()
+
+        return target_coordinate_system.transform_points(
+            self.site_coordinate_system, x_points, y_points)[:, 0:2]
+
+    @staticmethod
+    def check_sites_are_within_domain(sites, site_coords, site_x_coords,
+                                      site_y_coords, cube):
+        """
+        A function to remove sites from consideration if they fall outside the
+        domain of the provided model cube. A warning is raised and the details
+        of each rejected site are printed.
+
+        Args:
+            sites (list of dicts):
+                A list of dictionaries defining the spot sites for which
+                neighbours are to be found.
+            site_coords (np.array):
+                An array of shape (n_sites, 2) that contains the spot site
+                coordinates in the coordinate system of the model cube.
+            site_x_coords (np.array):
+                The x coordinates of the spot sites in their original
+                coordinate system, from which invalid sites must be removed.
+            site_y_coords (np.array):
+                The y coordinates of the spot sites in their original
+                coordinate system, from which invalid sites must be removed.
+            cube (iris.cube.Cube):
+                A cube that is representative of the model/grid from which spot
+                data will be extracted.
+        """
+        # Get the grid domain limits
+        x_min = cube.coord(axis='x').bounds[0][0]
+        x_max = cube.coord(axis='x').bounds[-1][1]
+        y_min = cube.coord(axis='y').bounds[0][0]
+        y_max = cube.coord(axis='y').bounds[-1][1]
+
+        domain_valid = np.where(
+            (site_coords[:, 0] >= x_min) & (site_coords[:, 0] <= x_max) &
+            (site_coords[:, 1] >= y_min) & (site_coords[:, 1] <= y_max))
+
+        domain_invalid = np.where(
+            (site_coords[:, 0] < x_min) | (site_coords[:, 0] > x_max) |
+            (site_coords[:, 1] < y_min) | (site_coords[:, 1] > y_max))
+
+        num_invalid = len(domain_invalid[0])
+        if num_invalid > 0:
+            msg = ("{} spot sites fall outside the grid domain and will not be"
+                   " processed. These sites are:\n".format(num_invalid))
+            dyn_msg = '{}\n'
+            for site in np.array(sites)[domain_invalid]:
+                msg = msg + dyn_msg.format(site)
+            warnings.warn(msg)
+
+        sites = np.array(sites)[domain_valid]
+        site_coords = site_coords[domain_valid]
+        site_x_coords = site_x_coords[domain_valid]
+        site_y_coords = site_y_coords[domain_valid]
+        return sites, site_coords, site_x_coords, site_y_coords
+
+    @staticmethod
+    def get_nearest_indices(site_coords, cube):
```

Comment:
It will return the nearest point from the domain, even if the site is well off the edge of it. That's why the check_sites_are_within_domain function is invoked first.